features/26-mfa/api.md

```diff
@@ -0,0 +1,436 @@
+# Overview
+
+The following factors are supported:
+
+- Time-Based One-Time Password [(TOTP)](https://tools.ietf.org/html/rfc6238)
+- Out-of-band
+  - SMS
+- Recovery code
+- Bearer token
+
+In the future, the below factors will be supported:
+
+- [WebAuthn with CTAP2 or FIDO U2F](https://fidoalliance.org/specifications/)
+
+The developer can configure the enforcement of MFA.
+
+The access token contains how the user was authenticated.
+
+# Configuration
+
+```yaml
+mfa:
+  # 'optional' or 'required'
+  #
+  # 'optional' means the user is free to register a new authenticator.
+  # When a user has at least one active authenticator, the user must login with MFA.
+  # This is like how other online services such as Google and Facebook offer MFA.
+  #
+  # 'required' means the user must login with MFA.
+  # APIs like change password will return MFA required error.
+  # This is like how banking services offer MFA.
+  #
+  # Default is 'optional'
```

Comment:
Alternatively, I think we need a configuration for the idea of "when mfa is needed if available", and the choices could be:

1. Always -- if the user have MFA, it is always required (default)
2. When not in IP address ranges
3. When country change
4. If metadata is set as X 

I think we can only support 1 and 2 to avoid complexity, but at least we have this concept in it?

Reply:
There are more rules we cannot think of. Case 2 and 3 can be done if the gateway inject `x-real-ip` and the gateway can infer country from IP address. I hope injecting informational headers can help application to perform their complex logic. However, the downside is that the API of auth gear is not affected by this mechanism.

Reply:
So....?

Reply:
The value `optional` corresponds to Case 1.

---

Case 4 is impossible now because metadata can be freely set by the user. We need to have labels/tags that can be set on users only by the developer to achieve this.

---

For Case 2, I am not sure about the actual cases and here are my guess.

If the request is not from whitelisted IP address ranges, MFA is required. So the user must associate at least one MFA factor.

If the request is from whitelisted IP address ranges, MFA is optional.

This situation corresponds to `required` but conditionally `optional` when the request is whitelisted.

It seems to me that the application is targeting regional users. Do you know any real world examples?

One use case to whitelist IP so that MFA can be skipped I can think of is programmatic usage.

Google has a feature called App Password. App Password is available only when the account has MFA turned on. Authentication with App Password does not require MFA. App Password can also be revoked at wish.

If we have Google App Password in Skygear, then it would be a kind of Identity Provider, just like password, OAuth and Custom Token. But it is a special case that does not require MFA.

---

For Case 3, I think it is about re-authentication.

When a user is created, we must remember the request IP and the inferred country. The user does not setup MFA and is able to login as long as the country is the same. The user travels to another country. MFA is required now and the user has to setup MFA. It seems unnatural to me that MFA setup is delayed until the user has traveled to another country. When the user travels back to the original country, MFA has been setup but is not required during login. This contradicts with how real world applications work.

A more natural way is to require re-authentication when the inferred country is different from the country when the access token was issued. The access token includes the request IP and the inferred country. The server compares the request country and the claimed country in the access token. If they are different, then re-authentication is required. The re-authentication can start from the first factor such as password or directly from the second factor like TOTP.

Reply:
@louischan-oursky oh I should describe more about the use cases.

For Case 2, the use cases could be for example, if the user login from a company's premises, it doesn't need 2FA. It could be very meaningful if the 2FA is a hardware dongle (say when we support WebAuthn)

For Case 3, it is not about re-auth. It is supposed for end-users who already have 2FA setup, but usually it is not needed. But if it is logging in from another country, we will require 2FA again. In fact I'm kind of randomly come up with this requirement, because ultimately I think this option should change to "Only require MFA for abnormal login", so we should combine multiple factors (AI-blah-blah-blah) of login context to requires 2FA login. Kind of like what Google Login normally do after you login the first time on acomputer.

For Case 4, maybe it is still okay if the metadata can be changed? It is for more complex logic like, for example in Oursky.com. If a user don't have 2FA enabled, he can only access Email, but not Calendar or Google Drive. After a user have 2FA enabled, he can access everything. So programatically, I think we can have a metadata like role, and only certain value will require 2FA, and the authorization code can check if mfa is present for certain access control.

(But of coz, if the developer doesn't want role is changeable by the user alone, he will need something else...)

======

Alternatively, I think we could have 3 options for this option:

`required` / `optional` / `smart` (only for abnormal login, but now our logic to detect abnormal could be dumb lol)